python/ray/tune/hpo_scheduler.py

```diff
@@ -0,0 +1,201 @@
+                config=new_cfg,
+                local_dir=os.path.join(self.args.local_dir, self._output_path),
+                experiment_tag=experiment_tag,
+                resources=self.args.resources,
+                stopping_criterion=self.args.stop,
+                checkpoint_freq=self.args.checkpoint_freq,
+                restore_path=self.args.restore,
+                upload_dir=self.args.upload_dir,
+                max_failures=self.args.max_failures)
+
+            self._tune_to_hp[trial] = new_trial_id
+            self._num_trials_left -= 1
+            yield trial
+
+    def on_trial_result(self, trial_runner, trial, result):
+        ho_trial = self._get_dynamic_trial(self._tune_to_hp[trial])
+        now = hpo.utils.coarse_utcnow()
+        ho_trial['book_time'] = now
+        ho_trial['refresh_time'] = now
+        return TrialScheduler.CONTINUE
+
+    def on_trial_error(self, trial_runner, trial):
+        ho_trial = self._get_dynamic_trial(self._tune_to_hp[trial])
+        ho_trial['refresh_time'] = hpo.utils.coarse_utcnow()
+        ho_trial['state'] = hpo.base.JOB_STATE_ERROR
+        ho_trial['misc']['error'] = (str(TuneError), "Tune Error")
+        self._hpopt_trials.refresh()
+        del self._tune_to_hp[trial]
+
+    def on_trial_remove(self, trial_runner, trial):
+        ho_trial = self._get_dynamic_trial(self._tune_to_hp[trial])
+        ho_trial['refresh_time'] = hpo.utils.coarse_utcnow()
+        ho_trial['state'] = hpo.base.JOB_STATE_ERROR
+        ho_trial['misc']['error'] = (str(TuneError), "Tune Removed")
+        self._hpopt_trials.refresh()
+        del self._tune_to_hp[trial]
+
+    def on_trial_complete(self, trial_runner, trial, result):
+        ho_trial = self._get_dynamic_trial(self._tune_to_hp[trial])
+        ho_trial['refresh_time'] = hpo.utils.coarse_utcnow()
+        ho_trial['state'] = hpo.base.JOB_STATE_DONE
+        hp_result = self._convert_result(result)
+        ho_trial['result'] = hp_result
+        self._hpopt_trials.refresh()
+        del self._tune_to_hp[trial]
+
+    def _convert_result(self, result):
+        return {"loss": getattr(result, self._loss_attr),
+                "status": "ok"}
+
+    def _get_dynamic_trial(self, tid):
+        return [t for t in self._hpopt_trials.trials if t["tid"] == tid][0]
+
+    def _continue(self):
+        return self._num_trials_left > 0
+
+    def get_hyperopt_trials(self):
+        return self._hpopt_trials
+
+    def choose_trial_to_run(self, trial_runner):
+        self.refresh_queue(trial_runner)
+        return FIFOScheduler.choose_trial_to_run(self, trial_runner)
+
+    def refresh_queue(self, trial_runner):
+        """Checks if there is a next trial ready to be queued.
+
+        This is determined by tracking the number of concurrent
+        experiments and trials left to run."""
+        while (self._num_trials_left > 0 and
+               self._num_live_trials() < self._max_concurrent):
```

Comment:
You can also check if there does not exist any PENDING trials. That means we can add another one.